x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/InferenceToXContentCompressor.java

```diff
@@ -82,32 +79,31 @@ private InferenceToXContentCompressor() {}
         }
     }
 
-    static Map<String, Object> inflateToMap(String compressedString) throws IOException {
+    static Map<String, Object> inflateToMap(BytesReference compressedBytes) throws IOException {
         // Don't need the xcontent registry as we are not deflating named objects.
         try(XContentParser parser = JsonXContent.jsonXContent.createParser(NamedXContentRegistry.EMPTY,
             LoggingDeprecationHandler.INSTANCE,
-            inflate(compressedString, MAX_INFLATED_BYTES))) {
+            inflate(compressedBytes, MAX_INFLATED_BYTES))) {
             return parser.mapOrdered();
         }
     }
 
-    static InputStream inflate(String compressedString, long streamSize) throws IOException {
-        byte[] compressedBytes = Base64.getDecoder().decode(compressedString.getBytes(StandardCharsets.UTF_8));
+    static InputStream inflate(BytesReference compressedBytes, long streamSize) throws IOException {
         // If the compressed length is already too large, it make sense that the inflated length would be as well
         // In the extremely small string case, the compressed data could actually be longer than the compressed stream
-        if (compressedBytes.length > Math.max(100L, streamSize)) {
+        if (compressedBytes.length() > Math.max(100L, streamSize)) {
             throw new CircuitBreakingException("compressed stream is longer than maximum allowed bytes [" + streamSize + "]",
                 CircuitBreaker.Durability.PERMANENT);
         }
-        InputStream gzipStream = new GZIPInputStream(new BytesArray(compressedBytes).streamInput(), BUFFER_SIZE);
+        InputStream gzipStream = new GZIPInputStream(compressedBytes.streamInput(), BUFFER_SIZE);
         return new SimpleBoundedInputStream(gzipStream, streamSize);
     }
 
-    private static String deflate(BytesReference reference) throws IOException {
+    private static BytesReference deflate(BytesReference reference) throws IOException {
         BytesStreamOutput out = new BytesStreamOutput();
         try (OutputStream compressedOutput = new GZIPOutputStream(out, BUFFER_SIZE)) {
             reference.writeTo(compressedOutput);
         }
-        return new String(Base64.getEncoder().encode(BytesReference.toBytes(out.bytes())), StandardCharsets.UTF_8);
```

Comment:
OK, so this will write out the raw bytes of the GZIP. Is this what we want or do we want to run the Base64 encoder? 

I thought the guarantees around base64 character sizes was one of the reasons we could skip transforming into a string?

Reply:
> Binary data is stored in lucene base64 encoded,

Ah, so since the mapping is `binary` we get that for free.

Reply:
It's handled by the Jackson JSON generator which is used by the various `XContentBuilder::value(byte[] value)` methods to write bytes